### ISSUE TITLE
sftp.makedirs

### ISSUE THREAD

**Issue by josephernest** [ronf/asyncssh#334]
Hello @ronf 
A useful SFTP function is `sftp.makedirs(...)` which creates the folders, sub, sub-subfolders, etc. if necessary (and does not raise an error if already exists). See for example https://pysftp.readthedocs.io/en/release_0.2.8/pysftp.html#pysftp.Connection.makedirs
Do you plan to add it? It would be handy :)

**Comment by josephernest** [comment]
This is not tested in detail, but would this work?

```
async def makedirs(self, remotedir):
    if await self.isdir(remotedir):  # already exists, it's ok
        return
    if await self.isfile(remotedir):
        print('A file already exists with the same name')  # or probably raise an exception?
    head, tail = os.path.split(remotedir)
    if head and not await self.isdir(head):
        await self.makedirs(head)
    if tail:
        await self.mkdir(remotedir)
```

A `rmtree` would be useful as well, to remove a directory and its files + subdirs etc. :) Did you plan to implement it?

**Comment by ronf** [comment]
This seems like a reasonable thing to consider adding. I'd probably want to implement it using an iterator rather than recursion, though. Also, when accessing remote paths, SFTP expects them to be sent as UTF-8 strings with '/' as the separator (more like UNIX paths), so it's not a good idea to use os.path. PurePath works well, though, and provides a "parts" property which does the splitting nicely. Then, it's just a matter of walking down the list and checking whether each entry already exists as a directory or not.

I don't think you need to do the regular file check -- if it's anything other than a directory, we probably want to return a failure of some kind, though we'd probably want it to follow symlinks so a part of the path could be a symlink to a directory elsewhere in the filesystem.

It would also be worth thinking about allowing the "exist_ok" argument that the regular os.makedirs() supports. We could also support "mode", though I don't believe SFTP does anything with the local umask, so the behavior might not exactly match what happens locally in that regard.

I've got some other work in progress right now, so I may not get to it right away, but thanks for the suggestion!

Adding `rmtree` also seems reasonable, though there are definitely some race conditions with that if anyone is touching any files in the subtree being removed, and SFTP doesn't provide the mechanisms used in the current Python local rmtree() to try and deal with that.

**Comment by ronf** [comment]
I got a chance to look at this today and have added a makedirs() method to SFTPClient in commit a0a1426, now available in the "develop" branch.

I ended up adding an "attrs" argument instead of "mode" to match what is done in the current mkdir() method. This lets you specify a mode using the "permissions" attribute, but it also allows you to set other attributes like UID & GID. I also added the exist_ok argument with behavior matching os.makedirs().

I'll take a look at adding rmtree() next.

Thanks for the suggestions!

**Comment by ronf** [comment]
An implementation of rmtree() is now available as commit fc9a06b in the "develop" branch. It takes the same arguments as shutil.rmtree() and supports the same error handling options. However, the specific functions passed into the error handlers will be bound method pointers on the SFTPClient object rather than local OS functions, and the exceptions will be SFTPError exceptions.

Removals are done in parallel to improve speed, and to facilitate this I added a new scandir() method which returns an async iterator of the SFTPName entries representing the contents of a directory, rather than collecting the full set of entries and returning them all at once. The prior readdir() and listdir() methods are still present, but they are now implemented using scandir().

**Comment by ronf** [comment]
Just checked in some improvements to rmtree() in commit d178dfb. It now limits the number of parallel scandir() requests allowed at once, and also better handles when multiple exceptions occur, avoiding possible "Future exception was never retrieved" error messages output by asyncio.

**Comment by ronf** [comment]
These changes are now released in AsyncSSH 2.5.0. Please let me know if you run into any problems with them.
